h2/src/main/org/h2/mvstore/db/LobStorageMap.java

```diff
@@ -380,18 +384,39 @@ public void removeAllForTable(int tableId) {
 
     @Override
     public void removeLob(ValueLob lob) {
+        LobDataDatabase lobData = (LobDataDatabase) lob.getLobData();
+        int tableId = lobData.getTableId();
+        long lobId = lobData.getLobId();
+        requestLobRemoval(tableId, lobId);
+    }
+
+    private void requestLobRemoval(int tableId, long lobId) {
+        pendingLobRemovals.offer(new LobRemovalInfo(mvStore.getCurrentVersion(), lobId, tableId));
+    }
+
+    @Override
+    public boolean needCleanup() {
+        return !pendingLobRemovals.isEmpty();
+    }
+
+    @Override
+    public void cleanup(long oldestVersionToKeep) {
         MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
         try {
-            LobDataDatabase lobData = (LobDataDatabase) lob.getLobData();
-            int tableId = lobData.getTableId();
-            long lobId = lobData.getLobId();
-            removeLob(tableId, lobId);
+            LobRemovalInfo lobRemovalInfo;
+            while ((lobRemovalInfo = pendingLobRemovals.poll()) != null && lobRemovalInfo.version < oldestVersionToKeep) {
```

Comment:
if have a transaction sequence that looks like:

    transaction 1: start
    transaction 2: start
    transaction 1: end
    .... very long time
    ... lots more transactions
    transaction 2: end

then because of this loop, the queue will get very big, and then we will have a major stall when all the LOBs actually get removed.

I think it would be better to have a queue that is sorted by version?

Reply:
Removal queue is sorted by version, just not explicitly, but naturally, by the order of insertions. 
And here I mean sorted by version, in which lob gets "deleted". By deletion I mean it becomes not reachable from current version of the database  / store. As for temporary lob, it gets deleted, when owning session closes At this point we put it into "pending removals" queue. It may still be reachable from a number of previous versions, and we have to wait until those versions will go out of scope and dropped, which will happen once all transactions, which were using those versions, are closed.
In your example we may end up with a huge queue indeed, but those lobs can not be removed at any earlier time, because they might still be in use. Of cause, "might" is a key word here, and it is possible to do a more granular reachability analysis, linking lobs to transaction(s), statement, result sets, but IMHO it's going to be even more expensive,
We have a long running transaction here, and this is going to be a problem for the whole database engine, not just LOB part. All store versions since transaction started will be retained, it's closure, so storage space may grow significantly.
At some point, restriction on transaction temporal length should be implemented, because it's rollback is a preferred outcome, compare to a risk of a database failure. Same can be said about LOB handling specifically. And it just now occurred to me that once queue mecomes unreasonably long we should start removing LOBs regardless of the version, until size of the queue will go below the limit. Majority of those lobs will be unreachable at this time anyway, but some may not, and at the point, where we currently getting NPE, exception should be thrown, saying that lob was held for too long.